vm/src/memory/heap.h

```diff
@@ -13,15 +13,42 @@ namespace msh {
     /**
      * A vector of heap allocated objects.
      */
-    struct obj_vector;
-    /**
-     * An object of the language that can be stored in the heap.
-     */
-    using obj = std::variant<int64_t, double, const std::string, obj_vector>;
+
+    class obj;
+
     struct obj_vector : public std::vector<obj *> {
         using std::vector<obj *>::vector;
     };
 
+    using obj_data = std::variant<int64_t, double, const std::string, obj_vector>;
+
+    class gc;
+
+    /**
+     * A vm object that can be stored in the heap.
+     */
+    class obj {
+        uint8_t gc_cycle;
+        obj_data data;
+
+        friend gc;
+
+    public:
+        template <typename T>
+        obj(T val) : gc_cycle{0}, data{val} {}
```

Comment:
```suggestion
        obj(obj_data val) : gc_cycle{}, data{std::move(val)} {}
```
A bit opinionated, but `std::variant` already has a non-explicit constructor. It does break code in `constant_pool.cpp` through.